proposals/NNNN-high-level-interface-modal-view-controllers.md

```diff
@@ -0,0 +1,160 @@
+
+## Motivation
+
+In order to work with SDL, app developers need to learn a new API which mostly doesn't adopt familiar patterns. Including but not limited to 
+- read best practices and implement proxy lifecycle (Android above all)
+- the use of RPCs and handle notifications, requests and responses
+- manually manage concurrent and sequential operations (e.g. image upload)
+
+The time for an app developer to learn SDL (non-productive time) is high and not accepted by some developers. The SDL iOS library already improves the current situation by abstracting painful implementations. However the management layer is still unfamiliar and causes a high learning curve.
+
+## Proposed solution
+
+This proposal is about adding modal view controllers to the SDL library which extends `SDLViewController` (see [SDL 0176](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0176-high-level-interface-views-and-controllers.md)) used by `SDLApplication` (see [SDL 0156](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0156-high-level-interface-foundation.md)). With modal view controllers app developers can use RPCs showing overlays (such as Alert or Slider) and structure the SDL related code.
+
+This proposal does not contain a modal representation of interacting choice sets. This will be proposed separately.
+
+### Modal view controllers
+
+Modal view controllers are responsible for overlay related RPCs such as `Alert`, `ScrollableMessage`, `PerformAudioPassThru`, `Slider` or `PerformInteraction` (proposed separately). Implementing overlays as modal view controllers would follow the modal-mode presentation of view controllers such as [`UIAlertController`](https://developer.apple.com/documentation/uikit/uialertcontroller).
+
+Apps can instanciate modal view controller classes and present them using the view controller manager (VC manager).
+
+#### SDLViewControllerManager (additions)
+
+```objc
+@interface SDLViewControllerManager
+
+/**
+ * This view controller is currently presenting a modal view on the screen.
+ */
+@property (nonatomic) SDLViewController *presentingModalViewController;
+
+/**
+ * This modal view controller is currently presented on the screen (by the presenting view controller).
+ */
+@property (nonatomic) SDLModalViewController *presentedModalViewController;
+
+/**
+ * Presents a modal view controller on the screen.
+ * Calls the completion handler if the modal view controller disappears from the screen or if an error occurred.
+ */
+- (void)presentModalViewController:(SDLModalViewController *)modalViewController completion:(void (^)(SDLResult result, NSError *error))completion;
```

Comment:
`CPInterfaceController.presentTemplate` is for non-modal UI. The API is more aligned with UIKit than with CarPlay.

Reply:
I believe it's actually only for modal UI. They have `pushTemplate`, and `popTemplate` for the non-modal case. What about `presentModal`?

Additionally, there has been discussion on the SDLC steering committee calls about explicitly not conforming these types of what we call modal behaviors as modals. During core certification guideline review, mention of "pop-up" and "modal" was explicitly criticized and removed due to some steering committee members not wanting to be restricted into making those objects modal. I'm wondering if we should rename this, but I'm not sure what it would be renamed to. Perhaps these could simply be `SDLView`s and the primary layouts `SDLTemplate`s?